x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeComputeHandler.java

```diff
@@ -342,15 +342,16 @@ private void acquireSearchContexts(
                         // we need to limit the number of active search contexts here or in SearchService
                         context = searchService.createSearchContext(shardRequest, SearchService.NO_TIMEOUT);
                         context.preProcess();
-                        searchContexts.add(context);
                     } catch (Exception e) {
                         if (addShardLevelFailure(shard.shardId(), e)) {
                             IOUtils.close(context);
+                            continue;
                         } else {
                             IOUtils.closeWhileHandlingException(context, () -> IOUtils.close(searchContexts));
                             throw e;
                         }
                     }
+                    searchContexts.add(context);
```

Comment:
NIT: I believe it is a bit more readable if context is added in the end of the try block as this emphasizes it is only added when operation is completely successful. Otherwise it looks like a null could be added to a collection sometimes.

Reply:
Sure, I've updated this in https://github.com/elastic/elasticsearch/pull/128222/commits/6a8994700a678330dfa7dfd672d0e2417da81152